### PR TITLE
change on location pipeline

### PR DESCRIPTION
**Pull request by Jaacky-Huaang**

### CODE REVIEW

File: src/pipeline.py [hunk 1]
```diff
@@ -160,29 +161,44 @@ def generate_grid(locations):
 
 def get_location_descriptors(mixvpr_agg, cpip_model, locations):
     location_encoder = cpip_model.location_projection
-    with torch.no_grad():
-        encoded_locations = location_encoder(locations)  # Encode all location vectors
-        # let num_points = ((2*CFG.grid_extent+1)**2 - 1) * CFG.num_rotation_steps,
-        # encoded_locations shape: (b, num_points, CFG.contrastive_dimension)
+    cnn_model = cpip_model.cnn
+    dotproduct_model = cpip_model.dotproduct
```
Comment (leo4life2) [inline src/pipeline.py]: not used anywhere?

File: src/pipeline.py [hunk 2]
```diff
@@ -160,29 +161,44 @@ def generate_grid(locations):
...
-        descriptors = cnn_model(reshaped_locations)
+            # current contrastive_dimension = 256
+            # Reshape to [1, sqrt(256), sqrt(256)] for each vector
+            b, num_points, _ = encoded_locations.shape
+            dimension_size = int(CFG.contrastive_dimension**0.5)
+            reshaped_locations = encoded_locations.view(b * num_points, 1, dimension_size, dimension_size)# [24000, 1, 16, 16] 
+            
+            descriptors = cnn_model(reshaped_locations) 
+            # descriptors: [12000, 1024, 20, 20]
+            pdb.set_trace()
```
Comment (leo4life2) [inline src/pipeline.py]: let's remove all pdb related lines of code
i suggest using `python -m debugpy --listen 5678 --wait-for-client your_python_file.py args` to use the debugger so you don't need to put debugger related code in the files. this works with the vscode remote attach debugger ui.

File: src/modules.py [hunk 3]
```diff
@@ -11,13 +11,23 @@ class ShallowConvNet(nn.Module):
     def __init__(self, input_c, input_h, input_w, output_c, output_h, output_w):
         super(ShallowConvNet, self).__init__()
         self.conv1 = nn.Conv2d(input_c, output_c, kernel_size=3, padding=1)
-        self.adaptive_pool = nn.AdaptiveAvgPool2d((output_h, output_w))
+        self.adaptive_pool = nn.AdaptiveMaxPool2d((output_h, output_w))
```
Comment (leo4life2) [inline src/modules.py]: why?

File: src/modules.py [hunk 4]
```diff
@@ -11,13 +11,23 @@ class ShallowConvNet(nn.Module):
...
     def forward(self, x):
         x = self.conv1(x)
         x = self.adaptive_pool(x)
         return x
 
+# A new class for dot product model
+class DotProductModel(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(DotProductModel, self).__init__()
+        self.matrix = nn.Parameter(torch.randn(input_dim, output_dim))
```
Comment (leo4life2) [inline src/modules.py]: why use randn?

File: src/modules.py [hunk 5]
```diff
@@ -11,13 +11,23 @@ class ShallowConvNet(nn.Module):
...
-        self.adaptive_pool = nn.AdaptiveAvgPool2d((output_h, output_w))
+        self.adaptive_pool = nn.AdaptiveMaxPool2d((output_h, output_w))
 
     def forward(self, x):
         x = self.conv1(x)
         x = self.adaptive_pool(x)
         return x
 
+# A new class for dot product model
+class DotProductModel(nn.Module):
```
Comment (leo4life2) [inline src/modules.py]: to scale our flattened cnn outputs down to cpip level, i think we need to use an mlp (projectionhead), or else information would be lost.

**Review by leo4life2 — CHANGES_REQUESTED** [review]

File: src/modules.py (hunk 3 above)
Comment (Jaacky-Huaang) [reply src/modules.py]: Sorry, I was trying to find out of a bug, which turned out to be coda out of memory

File: src/pipeline.py (hunk 1 above)
Comment (Jaacky-Huaang) [reply src/pipeline.py]: sorry I don't quite get what you mean

**Review by leo4life2 — APPROVED** [review]

**Pull request merged**